src/index.ts

```diff
@@ -0,0 +1,76 @@
+import { Application, IApplicationOptions, IApplicationPlugin } from '@pixi/app';
+import { Group, Layer } from '@pixi/layers';
+import { diffuseGroup, lightGroup, normalGroup } from 'pixi-lights';
+
+import { Container } from '@pixi/display';
+import { ShadowFilter } from './filters/ShadowFilter';
+import { Sprite } from '@pixi/sprite';
+import { augmentApplication } from './mixins/Application';
+import { augmentContainer } from './mixins/Container';
+
+export { filterFuncs } from './filters/FilterFuncs';
+export { ShadowFilter } from './filters/ShadowFilter';
+export { ShadowMaskFilter } from './filters/ShadowMaskFilter';
+export { augmentApplication } from './mixins/Application';
+export { augmentContainer } from './mixins/Container';
+export { Shadow } from './Shadow';
+
+export interface ShadowsOptions {
+    useLights: boolean;
+}
+export class Shadows {
+    // The objects that will cast shadows
+    casterGroup = new Group();
+    casterLayer = new Layer(this.casterGroup);
+    // The objects that will remain ontop of the shadows
+    overlayGroup = new Group();
+    overlayLayer = new Layer(this.overlayGroup);
+    filter: ShadowFilter;
+    container = new Container();
+    diffuseLayer: Layer | undefined;
+    normalLayer: Layer | undefined;
+    lightLayer: Layer | undefined;
+    constructor(app: Application, options?: ShadowsOptions) {
+        // // Create the shadow filter
+        this.filter = new ShadowFilter(app.renderer.width, app.renderer.height);
+        // Set up the container mixin so that it tells the filter about the available shadows and objects
+        augmentContainer(this.casterGroup, this.overlayGroup, this.filter);
+        // Overwrite the application render method
+        augmentApplication(app, this.filter);
+        app.stage.addChild(this.container);
+        // Set up the shadow layers
+        app.stage.addChild(this.casterLayer, this.overlayLayer);
+        if (options?.useLights) {
+            // Set up pixi-light's layers
+            this.diffuseLayer = new Layer(diffuseGroup);
+            this.normalLayer = new Layer(normalGroup);
+            this.lightLayer = new Layer(lightGroup);
+            const diffuseBlackSprite = new Sprite(this.diffuseLayer.getRenderTexture());
+
+            diffuseBlackSprite.tint = 0;
+
+            app.stage.addChild(this.diffuseLayer, diffuseBlackSprite, this.normalLayer, this.lightLayer);
+            // Add the shadow filter to the diffuse layer
+            this.diffuseLayer.filters = [this.filter];
+        } else {
+            this.container.filters = [this.filter];
+        }
```

Comment:
```suggestion
        if (options?.useLights) {
            app.stage = new LayerStage();

            // Set up pixi-light's layers
            this.diffuseLayer = new Layer(diffuseGroup);
            this.normalLayer = new Layer(normalGroup);
            this.lightLayer = new Layer(lightGroup);
            const diffuseBlackSprite = new Sprite(this.diffuseLayer.getRenderTexture());

            diffuseBlackSprite.tint = 0;

            // Set up the shadow layers
            app.stage.addChild(this.casterLayer, this.overlayLayer);
            app.stage.addChild(this.container);

            // Set up the lighting layers
            app.stage.addChild(this.diffuseLayer, diffuseBlackSprite, this.normalLayer, this.lightLayer);
            // Add the shadow filter to the diffuse layer
            this.diffuseLayer.filters = [this.filter];
        } else {
            // Set up the shadow layers
            app.stage.addChild(this.casterLayer, this.overlayLayer);
            app.stage.addChild(this.container);
            this.container.filters = [this.filter];
        }
```

For pixi-lights to work in this version, the stage from pixi-layers should be used:
```ts
import { Group, Layer, Stage as LayerStage } from '@pixi/layers';
```

Reply:
I moved the stage replacement to `augmentApplication()`. so what's left here is reversing the order of `app.stage.addChild()` calls, adding the container after the layers, instead of before.
Also, the order does not change between the two branches (the `if` and `else`) so I think it's better to write it once, before the `if`. WDYT?

Reply:
For the splitting of logic, I am fine either way. I was on the fence whether splitting into several if statements was the better call, or keep it with duplicate code so it's slightly easier to read. Also considered creating a function for repeated code, and invoking it.